File: src/voxicity/ChunkID.java
/*
 * Copyright 2011, Erik Lund
 *
 * This file is part of Voxicity.
 *
 *  Voxicity is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Voxicity is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Voxicity.  If not, see <http://www.gnu.org/licenses/>.
 */

package voxicity;

import org.lwjgl.util.vector.Vector3f;

public class ChunkID
{
	final int x;
	final int y;
	final int z;

	// Takes global coords and creates chunk coords
	public ChunkID( int x, int y, int z )
	{
		int[] c = Coord.GlobalToChunkBase( x, y, z );
		this.x = c[0];
		this.y = c[1];
		this.z = c[2];
	}

	public ChunkID( ChunkID id )
	{
		this( id.x, id.y, id.z );
	}

	public Vector3f coords()
	{
		return new Vector3f( x, y, z );
	}

	public boolean equals( Object o )
	{
		if ( o == null )
			return false;

		if ( !(o instanceof ChunkID ) )
			return false;

		ChunkID c = (ChunkID)o;

		if ( this.x != c.x || this.y != c.y || this.z != c.z )
			return false;

		return true;
	}

	public int hashCode()
	{
		return x ^ y ^ z;
	}

	public ChunkID get( Direction d )
	{
		switch ( d )
		{
			case East:
				return new ChunkID( x + Constants.Chunk.side_length, y, z );
			case West:
				return new ChunkID( x - Constants.Chunk.side_length, y, z );
			case North:
				return new ChunkID( x, y, z + Constants.Chunk.side_length );
			case South:
				return new ChunkID( x, y, z - Constants.Chunk.side_length );
			case Up:
				return new ChunkID( x, y + Constants.Chunk.side_length, z );
			case Down:
				return new ChunkID( x, y - Constants.Chunk.side_length, z );
			default:
				return new ChunkID( this );
		}
	}
}

File: src/chunkid.cpp
#include "chunkid.h"

#include "coord.h"

#include <cmath>

ChunkID::ChunkID( int _x, int _y, int _z )
{
    auto coords = Coord::GlobalToChunkBase( _x, _y, _z );
    x = coords[0];
    y = coords[1];
    z = coords[2];
}

ChunkID::ChunkID( float x, float y, float z ):
    ChunkID( static_cast<int>(std::lrint(x)),
            static_cast<int>(std::lrint(y)),
            static_cast<int>(std::lrint(z)) )
{}

std::vector<int> ChunkID::coords() const
{
    return { x, y, z };
}

ChunkID ChunkID::get( Direction d ) const
{
    using namespace Constants::Chunk;

    switch( d )
    {
        case East:  return ChunkID( x + side_length, y, z );
        case West:  return ChunkID( x - side_length, y, z );
        case Up:    return ChunkID( x, y + side_length, z );
        case Down:  return ChunkID( x, y - side_length, z );
        case North: return ChunkID( x, y, z + side_length );
        case South: return ChunkID( x, y, z - side_length );
        default: return *this;
    }
}
